File: src/scanner.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
)

func dirSize(path string) (uint64, error) {
	var sum uint64 = 0
	var subdirSum uint64 = 0
	var files []os.FileInfo
	var err error

	files, err = ioutil.ReadDir(path)

	for _, f := range files {
		if f.IsDir() && f.Name() != "." && f.Name() != ".." {
			subdirSum, err = dirSize(path + "/" + f.Name())
			sum += uint64(subdirSum)
		} else {
			sum += uint64(f.Size())
		}
	}

	return sum, err
}

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <path>\n", os.Args[0])
		os.Exit(1)
	}

	var path = os.Args[1]
	var err error
	var sum uint64

	fmt.Printf("Scanning dir (%s)\n", path)

	sum, err = dirSize(path)

	if err != nil {
		fmt.Println(err)
		fmt.Printf("Failed. %d\n", sum)
		os.Exit(1)
	}

	fmt.Printf("Done. %d\n", sum)
	os.Exit(0)
}

File: src/scanner.rs
use std::fs::read_dir;
use std::path::Path;
use std::env;
use std::process::ExitCode;

fn dir_size(dirname: &Path) -> Result<u64, &'static str> {
  let mut sum: u64 = 0;

  if let Ok(entries) = read_dir(dirname) {
    for entry in entries {
      let entry = entry.unwrap();
      let path = entry.path();

      if path.is_dir() && !entry.file_name().into_string().unwrap().starts_with(".") {
        eprintln!("dir_size(): Recursing into dir. ({})", path.display());
        sum += dir_size(&path).unwrap();
      } else {
        if let Ok(metadata) = entry.metadata() {
          eprintln!("dir_size(): File size ({}): ({})",
            path.display(),
            metadata.len(),
          );
          sum += metadata.len();
        }

      }
    }
  }

  return Ok(sum);
}


fn main() -> ExitCode {

  println!("Path size scanner.");

  let args: Vec<String> = env::args().collect();
  // dbg!(args);
  if args.len() < 2 {
    eprintln!("Provide a filesystem path as first argument.");
    std::process::exit(1);
  }

  let file_path = &args[1];

  if !Path::new(file_path).is_dir() {
    eprintln!("Not a valid directory: {}", file_path);
    std::process::exit(3);
  }

  println!("Scanning {}", file_path);
  let result = dir_size(Path::new(file_path));

  match result {
    Ok(_) => {
      println!("Done. {}", result.unwrap());
      std::process::exit(0);
    }
    Err(e) => {
      eprintln!("Failed: {}", e);
      std::process::exit(2);
    }
  }
}
